### PR TITLE
Add "desynchronized" flag to WebGLContextAttributes.

### PR DESCRIPTION
**Pull request by kenrussell**
Developers of stylus-based drawing applications (e.g. Google Keep)
have found this hint to be critical in order to be competitive with
native applications. It has been prototyped in Chromium and yields
significant latency improvements there.

This is the WebGL analogue of https://github.com/whatwg/html/pull/4360
and partially addresses https://github.com/whatwg/html/issues/4087 .

Principally authored by @yellowdoge, Miguel Casas-Sanchez <mcasas@chromium.org>

### CODE REVIEW

**Comment by kenrussell** [comment]
Mozilla / Apple / Microsoft folks in particular: please review.

@yellowdoge @fserb @wffurr FYI.

File: specs/latest/1.0/index.html [hunk 1]
```diff
@@ -744,6 +744,7 @@ <h3><a name="WEBGLCONTEXTATTRIBUTES">WebGLContextAttributes</a></h3>
     GLboolean preserveDrawingBuffer = false;
     WebGLPowerPreference powerPreference = "default";
     GLboolean failIfMajorPerformanceCaveat = false;
+    GLboolean lowLatency = false;
```
Comment (jdashg) [inline specs/latest/1.0/index.html]: Hah, why on earth are these GLbooleans?

File: specs/latest/1.0/index.html [hunk 2]
```diff
@@ -877,6 +878,21 @@ <h4>Context creation parameters</h4>
               set to <code>false</code>, with the knowledge that a reduced-fidelity rendering mode
               should be used to improve performance.
             </dd>
+        <dt> <span class=prop-value>lowLatency</span></dt>
+            <dd>
+              <p>
+                If the value is true, the user agent will attempt to optimize the rendering of the
+                canvas to reduce latency, including potentially rendering from the front buffer.
+                This might desynchronize the canvas's display from the rest of the page and induce
+                tearing on some displays.
```
Comment (jdashg) [inline specs/latest/1.0/index.html]: How does this interact with preserveDrawingBuffer:false?

File: specs/latest/1.0/index.html [hunk 3]
```diff
@@ -877,6 +878,21 @@ <h4>Context creation parameters</h4>
               set to <code>false</code>, with the knowledge that a reduced-fidelity rendering mode
               should be used to improve performance.
             </dd>
+        <dt> <span class=prop-value>lowLatency</span></dt>
+            <dd>
+              <p>
+                If the value is true, the user agent will attempt to optimize the rendering of the
+                canvas to reduce latency, including potentially rendering from the front buffer.
+                This might desynchronize the canvas's display from the rest of the page and induce
```
Comment (jdashg) [inline specs/latest/1.0/index.html]: s/might desynchronsize/desynchromizes/
s/the canvas's display/Presentation of the Canvas/

Comment (jdashg) [inline specs/latest/1.0/index.html]: What's the general proposed design here? There shouldn't be much else on the page moving, so this shouldn't be wildly different from a well-optimized same-frame compositing path.

File: specs/latest/1.0/webgl.idl [hunk 4]
```diff
@@ -106,7 +107,8 @@ typedef (ImageBitmap or
          ImageData or
          HTMLImageElement or
          HTMLCanvasElement or
-         HTMLVideoElement) TexImageSource;
+         HTMLVideoElement or
```
Comment (jdashg) [inline specs/latest/1.0/webgl.idl]: Split this into a different PR.

File: specs/latest/1.0/webgl.idl [hunk 5]
```diff
@@ -532,7 +534,8 @@ interface mixin WebGLRenderingContextBase
     const GLenum UNPACK_COLORSPACE_CONVERSION_WEBGL = 0x9243;
     const GLenum BROWSER_DEFAULT_WEBGL          = 0x9244;
 
-    [Exposed=Window] readonly attribute HTMLCanvasElement canvas;
+    [Exposed=Window] readonly attribute (HTMLCanvasElement or OffscreenCanvas) canvas;
+    [Exposed=Worker] readonly attribute OffscreenCanvas canvas;
```
Comment (jdashg) [inline specs/latest/1.0/webgl.idl]: Split this into a different PR.

File: specs/latest/1.0/index.html (hunk 2 above)
Comment (kenrussell) [reply specs/latest/1.0/index.html]: It's orthogonal. lowLatency:true can be used with preserveDrawingBuffer set to either true or false.

In Chromium, if lowLatency is true and, currently, if experimental command-line flags are set, then (on some OSs) WebGL's drawing buffer is allocated in GPU memory that targets the scanout hardware directly. Then, if preserveDrawingBuffer is true, the WebGL-rendered canvas still maintains an internal buffer for rendering, and presentation involves one blit to the scanout buffer. If preserveDrawingBuffer is false, draw calls go directly to the scanout buffer. (Actually, I'm not 100% sure whether single-, double- or triple-buffers are maintained for those scanout buffers.)

One drawing app, currently the primary requestor of this API, has put it into practice using preserveDrawingBuffer:true. They do only incremental rendering upon receiving new input events. The lowLatency flag, in desynchronizing from the rest of the page and getting to the display hardware sooner, yields dramatic latency improvements.

We aim to continue to experiment on various platforms and refine the spec text here. However, the current text is sufficient to allow the desired optimizations.

File: specs/latest/1.0/index.html (hunk 3 above)
Comment (kenrussell) [reply specs/latest/1.0/index.html]: We would prefer to leave desynchronization of the canvas's presentation optional. It seems likely to us that some browser vendors would not do this, but instead might change their internal buffering strategy somehow.

Comment (kenrussell) [reply specs/latest/1.0/index.html]: See the comment above for how this is currently implemented in Chromium. The use of the scanout buffer bypasses the browser's usual compositing pipeline and can break some guarantees about synchronization of content on the page, but does significantly reduce latency.

File: specs/latest/1.0/webgl.idl (hunk 4 above)
Comment (kenrussell) [reply specs/latest/1.0/webgl.idl]: Will do.

File: specs/latest/1.0/webgl.idl (hunk 5 above)
Comment (kenrussell) [reply specs/latest/1.0/webgl.idl]: Will do. Forgot to rerun the IDL generator after a recent commit.

File: specs/latest/1.0/index.html (hunk 3 above)
Comment (kenrussell) [reply specs/latest/1.0/index.html]: Note: in the current PR, have changed to use "presentation of the canvas".

**Comment by kenrussell** [comment]
Rebased on top of re-run of the IDL generator. Please re-review. Thanks.

**Comment by kenrussell** [comment]
Have rebased on top of other cleanup PRs.

This should be reviewed in conjunction with https://github.com/whatwg/html/issues/4087 and https://github.com/whatwg/html/pull/4234 , which describe adding the same flag for 2D canvas.

Note that a prototype of this flag has been tested in Chromium with some real-world drawing apps, and it significantly reduces latency. Hopefully that is enough motivation to spark other implementer interest. @jdashg @grorg @RafaelCintron

**Comment by kenrussell** [comment]
@mcasas note that I suggested to @jdashg , @grorg and @RafaelCintron to comment on https://github.com/whatwg/html/pull/4234 instead, because until that upstream PR is agreed upon, we can't add this to the WebGL specification.

**Comment by yellowdoge** [comment]
https://github.com/whatwg/html/pull/4360 (that superseded https://github.com/whatwg/html/pull/4234) has been merged, so we can adapt this PR (e.g. `s/lowLatency/desynchronized/`) and continue review, WDYT @kenrussell , @jdashg , @grorg and @RafaelCintron ?

**Comment by kenrussell** [comment]
@yellowdoge @jdashg this PR has been updated to use the same terminology as in the already-merged https://github.com/whatwg/html/pull/4360 . Updated the PR title and initial comment (as well as the commit's comment) to mirror it.

Could we please approve and merge this? Thanks.

**Comment by jdashg** [comment]
Thanks!
